pkg/sql/catalog/descs/collection.go

```diff
@@ -245,6 +251,57 @@ type Collection struct {
 	databaseLeasingUnsupported bool
 }
 
+type tempSchemaCacheKey struct {
+	dbID       descpb.ID
+	schemaName string
+}
+
+// getLeasedDescriptor return a leased descriptor valid for the transaction,
+// acquiring one if necessary. Due to a bug in lease acquisition for dropped
+// descriptors, the descriptor may have to be read from the store, in which case
+// shouldReadFromStore will be true.
+func (tc *Collection) getLeasedDescriptor(
+	ctx context.Context, txn *kv.Txn, parentID descpb.ID, parentSchemaID descpb.ID, name string,
+) (desc catalog.Descriptor, shouldReadFromStore bool, err error) {
+	// First, look to see if we already have the descriptor.
+	// This ensures that, once a SQL transaction resolved name N to id X, it will
+	// continue to use N to refer to X even if N is renamed during the
+	// transaction.
+	if desc = tc.leasedDescriptors.getByName(parentID, parentSchemaID, name); desc != nil {
+		log.VEventf(ctx, 2, "found descriptor in collection for '%s'", name)
+		return desc, false, nil
+	}
+
+	readTimestamp := txn.ReadTimestamp()
+	desc, expiration, err := tc.leaseMgr.AcquireByName(ctx, readTimestamp, parentID, parentSchemaID, name)
+	if err != nil {
+		// Read the descriptor from the store in the face of some specific errors
+		// because of a known limitation of AcquireByName. See the known
+		// limitations of AcquireByName for details.
+		if catalog.HasInactiveDescriptorError(err) ||
+			errors.Is(err, catalog.ErrDescriptorNotFound) {
+			return nil, true, nil
+		}
+		// Lease acquisition failed with some other error. This we don't
+		// know how to deal with, so propagate the error.
+		return nil, false, err
+	}
+
+	if expiration.LessEq(readTimestamp) {
+		log.Fatalf(ctx, "bad descriptor for T=%s, expiration=%s", readTimestamp, expiration)
+	}
+
+	tc.leasedDescriptors.add(desc)
+	log.VEventf(ctx, 2, "added descriptor '%s' to collection: %+v", name, desc)
+
+	// If the descriptor we just acquired expires before the txn's deadline,
+	// reduce the deadline. We use ReadTimestamp() that doesn't return the commit
+	// timestamp, so we need to set a deadline on the transaction to prevent it
+	// from committing beyond the version's expiration time.
+	txn.UpdateDeadlineMaybe(ctx, expiration)
```

Comment:
this method needs to take in the bool passed from higher levels about whether or not to actually call update

Reply:
Yeah, but in practice that's only needed when we're acquiring descriptors (specifically, types) by ID, and we haven't added that bool argument to any of the methods that resolve by name, which are the exclusive callers of this new method. I don't think it'd be useful to introduce that method argument now, especially since all the callers would pass true as the value. I renamed this function to `getLeasedDescriptorByName` to hopefully make things clearer.

Reply:
Thanks for the clarification, that sounds right.